apps/astarte_realm_management/lib/astarte_realm_management/engine.ex

```diff
@@ -937,6 +941,25 @@ defmodule Astarte.RealmManagement.Engine do
     end
   end
 
+  defp verify_mappings_max_storage_retention(realm_name, interface) do
+    with {:ok, max_retention} <- get_datastream_maximum_storage_retention(realm_name),
+         {:mappings_retention_valid, true} <-
+           {:mappings_retention_valid,
+            mappings_retention_valid?(interface.mappings, max_retention)} do
+      :ok
+    else
+      {:mappings_retention_valid, false} -> {:error, :maximum_database_retention_exceeded}
```

Comment:
Just do
```elixir
defp verify_mappings_max_storage_retention(realm_name, interface) do
  with {:ok, max_retention} <- get_datastream_maximum_storage_retention(realm_name) do
    if mappings_retention_valid?(interface.mappings, max_retention) do
      :ok
    else
      {:error, :maximum_database_retention_exceeded}
    end
  end
end
```

Some reasons why:
- It's better having `with` without `else`
- Tuple-tagging the rhs in `with` blocks seemed a good idea at the time so you'll find code like this around the codebase, but it's actually not and it's less readable than the `if` version

The other possibility is changing the validation function to return `:ok/{:error, :maximum_database_retention_exceeded}` instead of a boolean (usually that kind of function is then called `validate_*`, so `validate_mappings_retention`), then you can just do

```elixir
defp verify_mappings_max_storage_retention(realm_name, interface) do
  with {:ok, max_retention} <- get_datastream_maximum_storage_retention(realm_name) do
    validate_mappings_retention(interface.mappings, max_retention)
  end
end
```